clients/drcachesim/tests/analysis_unit_tests.cpp

```diff
@@ -177,10 +183,184 @@ test_queries()
+        bool
+        print_results() override
+        {
+            return true;
+        }
+        bool
+        parallel_shard_supported() override
+        {
+            return true;
+        }
+        void *
+        parallel_shard_init_stream(int shard_index, void *worker_data,
+                                   memtrace_stream_t *stream) override
+        {
+            auto per_shard = new per_shard_t;
+            per_shard->index = shard_index;
+            per_shard->stream = stream;
+            return reinterpret_cast<void *>(per_shard);
+        }
+        bool
+        parallel_shard_exit(void *shard_data) override
+        {
+            per_shard_t *shard = reinterpret_cast<per_shard_t *>(shard_data);
+            (*schedule_strings_)[shard->index] = shard->schedule;
+            delete shard;
+            return true;
+        }
+        bool
+        parallel_shard_memref(void *shard_data, const memref_t &memref) override
+        {
+            per_shard_t *shard = reinterpret_cast<per_shard_t *>(shard_data);
+            // We run in *rough* lockstep to avoid a flaky test: we just need to
+            // avoid the 2nd output making it through several initial records
+            // before the 1st output runs and sees a STATUS_WAIT.
+            static constexpr int MAX_WAITS = 100000;
+            int waits;
+            while (global_records_ < 3 * shard->records / 2) {
+                std::this_thread::yield();
+                // Avoid a hang.  It shouldn't happen with these inputs though.
+                if (++waits > MAX_WAITS)
+                    break;
+            }
+            ++shard->records;
+            ++global_records_;
+            if (memref.marker.type == TRACE_TYPE_MARKER &&
+                memref.marker.marker_type == TRACE_MARKER_TYPE_WAIT) {
+                shard->schedule += '-';
+                return true;
+            }
+            int64_t input = shard->stream->get_input_id();
+            shard->schedule += 'A' + static_cast<char>(input % 26);
+            return true;
+        }
+
+    private:
+        struct per_shard_t {
+            int index;
+            memtrace_stream_t *stream;
+            std::string schedule;
+            int64_t records = 0;
+        };
+        std::atomic<int64_t> global_records_ { 0 };
+        std::vector<std::string> *schedule_strings_;
+    };
+
+    std::vector<std::string> schedule_strings(NUM_OUTPUTS);
+    std::vector<analysis_tool_t *> tools;
+    auto test_tool = std::unique_ptr<test_tool_t>(new test_tool_t(&schedule_strings));
+    tools.push_back(test_tool.get());
+    mock_analyzer_t analyzer(sched_inputs, &tools[0], (int)tools.size(),
+                             /*parallel=*/true, NUM_OUTPUTS, &sched_ops);
+    assert(!!analyzer);
+    bool res = analyzer.run();
+    assert(res);
+    for (const auto &sched : schedule_strings) {
+        std::cerr << "Schedule: " << sched << "\n";
+    }
+    assert(schedule_strings[1][0] == '-');
```

Comment:
This is the final check the test does? That there was one wait? Add a comment.

Reply:
Added a comment about non-determinism (hard to make it purely lockstep from here) meaning we can't have a very strict check.